numba/array_analysis.py

```diff
@@ -41,846 +52,1266 @@
+                return cgutils.unpack_tuple(builder, ary.shape)
+            elif isinstance(aty, types.BaseTuple):
+                return cgutils.unpack_tuple(builder, a)
+            else: # otherwise it is a single integer
+                return [a]
+
+        def pairwise(a, aty, b, bty):
+            ashapes = unpack_shapes(a, aty)
+            bshapes = unpack_shapes(b, bty)
+            assert len(ashapes) == len(bshapes)
+            for (m, n) in zip(ashapes, bshapes):
+                m_eq_n = builder.icmp(lc.ICMP_EQ, m, n)
+                with builder.if_else(m_eq_n) as (then, orelse):
+                    with then:
+                        pass
+                    with orelse:
+                        context.call_conv.return_user_exc(
+                            builder, AssertionError, (msg,))
+
+        for i in range(1, len(tup_type) - 1):
+            pairwise(tup[i], tup_type[i], tup[i+1], tup_type[i+1])
+        r = context.get_constant_generic(builder, types.NoneType, None)
+        return r
+    return signature(types.none, *val), codegen
+
+
+class EquivSet(object):
+    """EquivSet keeps track of equivalence relations between
+    a set of objects.
+    """
+    def __init__(self, obj_to_ind = None, ind_to_obj = None, next_ind = 0):
+        if obj_to_ind == None:
+            obj_to_ind = {}
+        if ind_to_obj == None:
+            ind_to_obj = {}
+        self.obj_to_ind = obj_to_ind # object to index mapping
+        self.ind_to_obj = ind_to_obj # index to objects mapping
+        self.next_ind = next_ind
+
+    def empty(self):
+        """Return an empty EquivSet object.
+        """
+        return EquivSet()
+
+    def clone(self):
+        """Return a new copy.
+        """
+        return EquivSet(obj_to_ind = copy.deepcopy(self.obj_to_ind),
+                        ind_to_obj = copy.deepcopy(self.ind_to_obj),
+                        next_id = self.next_ind)
+
+    def __repr__(self):
+        return "EquivSet({})".format(self.ind_to_obj)
+
+    def is_empty(self):
+        """Return true if the set is empty, or false otherwise.
+        """
+        return len(self.obj_to_ind) == 0
+
+    def _get_ind(self, x):
+        """Return the internal index (greater or equal to 0) of the given
+        object, or -1 if not found.
+        """
+        if x in self.obj_to_ind:
+            return self.obj_to_ind[x]
+        return -1
+
+    def _get_or_add_ind(self, x):
+        """Return the internal index (greater or equal to 0) of the given
+        object, or create a new one if not found.
+        """
+        if x in self.obj_to_ind:
+            i = self.obj_to_ind[x]
+        else:
+            i = self.next_ind
+            self.next_ind += 1
+        return i
+
+    def _insert(self, objs):
+        """Base method that inserts a set of equivalent objects by modifying
+        self.
+        """
+        assert len(objs) > 1, "insert_equiv needs more than one argument"
```

Comment:
It is something that should never happen. Think I'll just remove the error message to avoid confusion.